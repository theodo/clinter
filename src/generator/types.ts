import { Linter } from "eslint";
import { ProjectInfoObject } from "types";

export type ESLintGenerator = (userAnswers: ProjectInfoObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: ProjectInfoObject) => (deps: string[]) => string[];

export interface ESLintOverrider {
  files: string[];
}
