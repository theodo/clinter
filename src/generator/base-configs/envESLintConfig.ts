import { Linter } from "eslint";

export const browserESLintEnvConfig: Linter.Config = {
  env: {
    browser: true,
    es6: true,
  },
};

export const nodeESLintEnvConfig: Linter.Config = {
  env: {
    node: true,
    es6: true,
  },
};

export const jestESLintEnvConfig: Linter.Config = {
  env: {
    jest: true,
  },
};

export const jestESLintDependencies = ["eslint-plugin-jest"];
