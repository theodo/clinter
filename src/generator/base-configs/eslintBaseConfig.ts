import { Linter } from "eslint";

export const eslintBaseConfig: Linter.Config = {
  root: true,
  extends: ["eslint:recommended"],
  plugins: ["import"],
  rules: {
    curly: ["error", "all"],
    eqeqeq: ["error", "smart"],
    "import/no-extraneous-dependencies": [
      "error",
      { devDependencies: true, optionalDependencies: false, peerDependencies: false },
    ],
    "no-shadow": [
      "error",
      {
        hoist: "all",
      },
    ],
    "prefer-const": "error",
    "import/order": [
      "error",
      {
        groups: [["external", "builtin"], "internal", ["parent", "sibling", "index"]],
      },
    ],
    "sort-imports": [
      "error",
      {
        ignoreCase: true,
        ignoreDeclarationSort: true,
        ignoreMemberSort: false,
        memberSyntaxSortOrder: ["none", "all", "multiple", "single"],
      },
    ],
    "padding-line-between-statements": ["error", { blankLine: "always", prev: "*", next: "return" }],
  },
};

export const eslintBaseDependencies = ["eslint", "eslint-plugin-import"];
