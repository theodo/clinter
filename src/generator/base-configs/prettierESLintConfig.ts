import { Linter } from "eslint";

export const prettierESLintConfig: Linter.Config = {
  extends: ["plugin:prettier/recommended"],
};

export const prettierESLintDependencies = ["prettier", "eslint-config-prettier", "eslint-plugin-prettier"];
