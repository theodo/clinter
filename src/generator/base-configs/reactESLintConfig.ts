import { Linter } from "eslint";

export const reactESLintConfig: Linter.Config = {
  settings: {
    react: {
      version: "detect",
    },
  },
  extends: ["react-app", "plugin:jsx-a11y/recommended"],
  rules: {
    "react/no-string-refs": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "warn",
  },
  plugins: ["jsx-a11y"],
};

export const reactTypescriptESLintConfig: Linter.Config = {
  rules: {
    "@typescript-eslint/explicit-function-return-type": "off",
  },
};

export const reactESLintDependencies = ["eslint-plugin-react-app", "eslint-config-react-app", "eslint-plugin-jsx-a11y"];
