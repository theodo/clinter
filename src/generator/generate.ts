import { Linter } from "eslint";
import { generateTestESLintConfig, getTestESLintDependencies } from "generator/test-framework";
import { generateTypescriptESLintConfig, getTypescriptESLintDependencies } from "generator/typescript";
import { generateEnvESLintConfig } from "generator/env";
import { generateFrontFrameworkESLintConfig, getFrontFrameworkESLintDependencies } from "generator/front-framework";
import { cleanPrettierConfig, generatePrettierESlintConfig, getPrettierESLintDependencies } from "generator/prettier";
import { eslintBaseConfig, eslintBaseDependencies } from "generator/base-configs/eslintBaseConfig";
import { ESLintDependencyGenerator, ESLintGenerator } from "generator/types";
import { areArraysEqual, mergeArrays, pipe } from "utils/utility";
import { ProjectInfoObject } from "types";

function addEslintOverride(
  overrides: Linter.ConfigOverride[],
  newOverride: Linter.ConfigOverride
): Linter.ConfigOverride[] {
  if (overrides.length === 0) {
    return [newOverride];
  }

  const [currentOverride, ...rest] = overrides;

  // Configuration overrides are entirely defined by their files property
  const areOverrideFilesEqual = areArraysEqual(
    typeof currentOverride.files === "string" ? [currentOverride.files] : currentOverride.files,
    typeof newOverride.files === "string" ? [newOverride.files] : newOverride.files
  );

  if (currentOverride.parser === newOverride.parser || areOverrideFilesEqual) {
    return [concatConfig(newOverride)(currentOverride), ...rest];
  }

  return [currentOverride, ...addEslintOverride(rest, newOverride)];
}

function concatEslintOverrides(
  prevOverrides: Linter.ConfigOverride[],
  nextOverrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  return nextOverrides.reduce(addEslintOverride, prevOverrides);
}

function concatESlintObjects<T extends Record<string, unknown>>(prevObject: T, nextObject: T): T {
  return { ...prevObject, ...nextObject };
}

function shouldKeepConfigEntry(configEntry: [string, unknown]): boolean {
  const [_, value] = configEntry;

  if (typeof value === "object" && value !== null && Object.keys(value).length === 0) {
    return false;
  }

  return true;
}

function cleanObjectConfig<T>(config: T): T {
  const entries = Object.entries(config);

  const cleanEntries = entries.filter(shouldKeepConfigEntry);

  return Object.fromEntries(cleanEntries) as T;
}

function cleanExtendsValue(
  extendsValue: string | string[],
  filterPredicate: (item: string) => boolean
): string | string[] {
  if (typeof extendsValue === "string") {
    return filterPredicate(extendsValue) ? extendsValue : [];
  } else {
    return extendsValue.filter(filterPredicate);
  }
}

function cleanExtendsField<T extends Linter.Config>(config: T, filterPredicate: (item: string) => boolean): T {
  const { extends: extendsValue } = config;
  if (extendsValue === undefined) {
    return config;
  }
  return {
    ...config,
    extends: cleanExtendsValue(extendsValue, filterPredicate),
  };
}

export function cleanESLintExtendsField(
  config: Linter.Config,
  filterPredicate: (item: string) => boolean
): Linter.Config {
  const { overrides, ...rest } = cleanExtendsField(config, filterPredicate);

  const cleanOverrides = overrides?.map((overrideConfig) => cleanExtendsField(overrideConfig, filterPredicate));

  if (overrides) {
    return { ...rest, overrides: cleanOverrides };
  }

  return { ...rest };
}

function clearEmptyFields(config: Linter.Config) {
  const { overrides, ...rest } = cleanObjectConfig(config);

  const cleanOverrides = overrides?.map(cleanObjectConfig);

  if (overrides) {
    return { ...rest, overrides: cleanOverrides };
  }

  return { ...rest };
}

function cleanESLintConfig(config: Linter.Config): Linter.Config {
  return pipe(cleanPrettierConfig, clearEmptyFields)(config);
}

function concatESlintArrays(prev: Array<string> | string, next: Array<string> | string): Array<string> {
  const prevArray = typeof prev === "string" ? [prev] : prev;
  const nextArray = typeof next === "string" ? [next] : next;

  return mergeArrays(prevArray, nextArray);
}

export function concatConfig<T extends Linter.Config | Linter.ConfigOverride>(config: T): (prevConfig: T) => T {
  return (prevConfig) => ({
    ...prevConfig,
    ...config,
    extends: concatESlintArrays(prevConfig.extends ?? [], config.extends ?? []),
    rules: concatESlintObjects(prevConfig.rules ?? {}, config.rules ?? {}),
    plugins: concatESlintArrays(prevConfig.plugins ?? [], config.plugins ?? []),
    parserOptions: concatESlintObjects(prevConfig.parserOptions ?? {}, config.parserOptions ?? {}),
    env: concatESlintObjects(prevConfig.env ?? {}, config.env ?? {}),
    overrides: concatEslintOverrides(prevConfig.overrides ?? [], config.overrides ?? []),
  });
}

export function concatDependencies(deps2: string[]): (deps: string[]) => string[] {
  return (deps1) => mergeArrays(deps1, deps2);
}

const generateBaseESLintConfig: ESLintGenerator = (_userAnswers) => {
  return concatConfig(eslintBaseConfig);
};

const getBaseESLintDependencies: ESLintDependencyGenerator = (_userAnswers) =>
  concatDependencies(eslintBaseDependencies);

export function generateEslintConfig(userAnswers: ProjectInfoObject, startConfig: Linter.Config = {}): Linter.Config {
  return pipe(
    generateBaseESLintConfig(userAnswers),
    generateTypescriptESLintConfig(userAnswers),
    generateEnvESLintConfig(userAnswers),
    generateTestESLintConfig(userAnswers),
    generateFrontFrameworkESLintConfig(userAnswers),
    // Prettier must be at the end of the list to avoid potential conflicts
    generatePrettierESlintConfig(userAnswers),
    cleanESLintConfig
  )(startConfig);
}

export function getConfigDependencies(userAnswers: ProjectInfoObject): string[] {
  return pipe(
    getBaseESLintDependencies(userAnswers),
    getTypescriptESLintDependencies(userAnswers),
    getTestESLintDependencies(userAnswers),
    getFrontFrameworkESLintDependencies(userAnswers),
    getPrettierESLintDependencies(userAnswers)
  )([]);
}
