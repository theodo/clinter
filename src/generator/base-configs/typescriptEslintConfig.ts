import { Linter } from "eslint";
import { ESLintOverrider } from "generator/types";

export const typescriptBaseEslintConfig: Linter.Config = {
  extends: ["plugin:@typescript-eslint/recommended"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "tsconfig.json",
  },
  rules: {
    // Prefer using concise optional chain expressions instead of chained logical ands
    "@typescript-eslint/prefer-optional-chain": "error",
    // Disable JS no-shadow rule
    "no-shadow": "off",
    // Disallow variable declarations from shadowing variables declared in the outer scope
    "@typescript-eslint/no-shadow": "error",
  },
};

export const typescriptTypeEslintConfig: Linter.Config = {
  extends: ["plugin:@typescript-eslint/recommended-requiring-type-checking"],
  rules: {
    // Enforce the usage of the nullish coalescing operator instead of logical chaining
    "@typescript-eslint/prefer-nullish-coalescing": "error",
    // Restricts the types allowed in boolean expressions
    "@typescript-eslint/strict-boolean-expressions": [
      "error",
      {
        allowString: false,
        allowNumber: false,
        allowNullableObject: true,
      },
    ],
    // Flags unnecessary equality comparisons against boolean literals
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
    // Prevents conditionals where the type is always truthy or always falsy
    "@typescript-eslint/no-unnecessary-condition": "error",
    // Enforces that type arguments will not be used if not required
    "@typescript-eslint/no-unnecessary-type-arguments": "error",
    // Enforce the use of String#startsWith and String#endsWith instead of other equivalent methods of checking substrings
    "@typescript-eslint/prefer-string-starts-ends-with": "error",
    // Exhaustiveness checking in switch with union type
    "@typescript-eslint/switch-exhaustiveness-check": "error",
    // Prevents the text 'undefined' to appear in a template string
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      {
        allowNumber: true,
        allowBoolean: true,
      },
    ],
  },
};

export const tsOverrider: ESLintOverrider = {
  files: ["**/*.ts?(x)"],
};

export const typescriptESLintDependencies = ["@typescript-eslint/parser", "@typescript-eslint/eslint-plugin"];
