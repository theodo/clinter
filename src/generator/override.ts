import { Linter } from "eslint";
import { ESLintOverrider } from "generator/types";

/**
 * Transform a config into an override config
 * @param overrider the eslint attributes to be added to transform the config to an override config
 */
export function wrapConfigInOverride(overrider: ESLintOverrider): (config: Linter.Config) => Linter.Config {
  return (config) => ({
    overrides: [
      {
        ...overrider,
        ...config,
      },
    ],
  });
}
