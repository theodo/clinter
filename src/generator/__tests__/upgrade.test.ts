import { generateEslintConfig } from "generator/generate";
import { FormatterInfo, FrontFrameworkInfo, TestFrameworkInfo, TypescriptInfo } from "types";
import { eslintBaseConfig } from "generator/base-configs";
import { Linter } from "eslint";

describe("generateEslintConfig", () => {
  describe("Upgrade Mode", () => {
    describe("Simple eslint config", () => {
      it("should return the base eslint config when no start config is given", () => {
        expect(
          generateEslintConfig(
            {
              typescript: TypescriptInfo.None,
              env: [],
              formatter: FormatterInfo.None,
              frontFramework: FrontFrameworkInfo.None,
              test: TestFrameworkInfo.None,
            },
            {}
          )
        ).toStrictEqual(eslintBaseConfig);
      });

      it("should return a ESLint config with empty fields removed", () => {
        expect(
          generateEslintConfig(
            {
              typescript: TypescriptInfo.None,
              env: [],
              formatter: FormatterInfo.None,
              frontFramework: FrontFrameworkInfo.None,
              test: TestFrameworkInfo.None,
            },
            {
              env: {},
            }
          )
        ).toStrictEqual(eslintBaseConfig);
      });

      it("should return a ESLint config without changing the order of the original config fields", () => {
        const customESLintConfig: Linter.Config = {
          env: {
            node: true,
            browser: true,
          },
          extends: ["CUSTOM-ESLINT-CONFIG"],
          root: true,
          rules: {
            "custom-rule-name": "error",
          },
        };
        expect(
          generateEslintConfig(
            {
              typescript: TypescriptInfo.None,
              env: [],
              formatter: FormatterInfo.None,
              frontFramework: FrontFrameworkInfo.None,
              test: TestFrameworkInfo.None,
            },
            customESLintConfig
          )
        ).toMatchSnapshot();
      });
    });
  });
});
