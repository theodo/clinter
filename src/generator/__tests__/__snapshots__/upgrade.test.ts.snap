// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEslintConfig Upgrade Mode Simple eslint config should return a ESLint config without changing the order of the original config fields 1`] = `
Object {
  "env": Object {
    "browser": true,
    "node": true,
  },
  "extends": Array [
    "CUSTOM-ESLINT-CONFIG",
    "eslint:recommended",
  ],
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "curly": Array [
      "error",
      "all",
    ],
    "custom-rule-name": "error",
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "padding-line-between-statements": Array [
      "error",
      Object {
        "blankLine": "always",
        "next": "return",
        "prev": "*",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;
